File: 300. Longest Increasing Subsequence/300(1).cpp
// intelligently build subsequence + binary search
// time: O(nlog(n))
// space: O(n)
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        // build increasing subsequence (not always valid subsequence but the length will always equal the length of the longest increasing subsequence)
        vector<int> subseq;
        for (int num: nums){
            // expand subsequence
            if (subseq.size() == 0 || num > subseq[subseq.size()- 1]){
                subseq.push_back(num);
            } else{
                // replace the first num that is greater than or equal to num
                // length is maintained but once I find a longer substring I can extend it 
                // can use binary search since it's sorted
                bool eqfound = false;
                int start = 0, end = subseq.size() - 1;
                while(start <= end){
                    int mid = start + (end-start) / 2;
                    if (subseq[mid] == num){
                        eqfound = true;
                        break;
                    } else if (num < subseq[mid]){
                        end = mid - 1;
                    } else{
                        start = mid + 1;
                    }
                }
                // if can't find element want to 
                if (!eqfound){
                    subseq[start] = num;
                }

            }
        }
        return subseq.size();
    }
};

File: 300. Longest Increasing Subsequence/300(1).py
"""
intelligently build increasing subsequence + binary search
time: O(n^2)
space: O(n)
"""
import bisect
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        if len(nums) ==0:
            return 0

        increasing_sub = []

        for num in nums:
            index = bisect.bisect_left(increasing_sub, num)

            # num is greater than all other elements in increasing_sub -> can include to build increasing_sub
            if index == len(increasing_sub):
                increasing_sub.append(num)
            else:
                increasing_sub[index] = num

        return len(increasing_sub)



"""
attempting to construct longest strictly increasing subsequence
motivation: 
- binary search has O(logn) optimize time complexity
- estrictly increasing subsequence is sorted

why increasing_sub[index] = num:
initial thought: only update when index == len(sub) - 1
can replace last element with smaller value while mainaining correct relative order
-> however think about this case: [3,5,6,2,5,4,19,5,6,7,12]
there might be future smaller starting potins that would lead to longer strictly increasing subsequence
-> update increasing_sub and once Im able to create a longer strictly increasing subsequence from it it will start to affect len(sub)

note that increasing_sub may not be the correct answer
"""
